File: eclipse-workspace/arduino_protocol/src/ArduinoStatus.cpp
/*
 * Arduino.cpp
 *
 *  Created on: 21/01/2015
 *      Author: cams7
 */

#include "ArduinoStatus.h"

namespace SISBARC {

const transmitter ArduinoStatus::ARDUINO = 0x00;
const transmitter ArduinoStatus::PC = 0x01;

const status ArduinoStatus::SEND = 0x00;
const status ArduinoStatus::SEND_RESPONSE = 0x01;
const status ArduinoStatus::RESPONSE = 0x02;
const status ArduinoStatus::RESPONSE_RESPONSE = 0x03;

const pin_type ArduinoStatus::DIGITAL = 0x00;
const pin_type ArduinoStatus::ANALOG = 0x01;

const uint8_t ArduinoStatus::PIN_MAX = 0x3F; //63
const uint16_t ArduinoStatus::PIN_VALUE_MAX = 0x3FF; //1023

const uint8_t ArduinoStatus::PINS_DIGITAL_SIZE = 6;
const uint8_t ArduinoStatus::PINS_DIGITAL[PINS_DIGITAL_SIZE] = { 2, 4, 7, 8, 12,
		13 };

const uint8_t ArduinoStatus::PINS_DIGITAL_PWM_SIZE = 6;
const uint8_t ArduinoStatus::PINS_DIGITAL_PWM[PINS_DIGITAL_PWM_SIZE] = { 3, 5,
		6, 9, 10, 11 };

const uint8_t ArduinoStatus::PINS_ANALOG_SIZE = 6;
const uint8_t ArduinoStatus::PINS_ANALOG[PINS_ANALOG_SIZE] =
		{ 0, 1, 2, 3, 4, 5 };

ArduinoStatus::ArduinoStatus() {
	setTransmitterValue(ARDUINO);

	setPinType(0x00);
	setPin(0x00);
	setPinValue(0x0000);

	setStatusValue(0x00);
}

ArduinoStatus::ArduinoStatus(pin_type pinType, uint8_t pin, uint16_t pinValue,
		status statusValue) {
	setTransmitterValue(ARDUINO);

	setPinType(pinType);
	setPin(pin);
	setPinValue(pinValue);

	setStatusValue(statusValue);
}

ArduinoStatus::~ArduinoStatus() {
	// TODO Auto-generated destructor stub
}

transmitter ArduinoStatus::getTransmitterValue(void) {
	return transmitterValue;
}

void ArduinoStatus::setTransmitterValue(transmitter transmitterValue) {
	this->transmitterValue = transmitterValue;
}

status ArduinoStatus::getStatusValue(void) {
	return statusValue;
}

void ArduinoStatus::setStatusValue(status statusValue) {
	this->statusValue = statusValue;
}

pin_type ArduinoStatus::getPinType(void) {
	return pinType;
}

void ArduinoStatus::setPinType(pin_type pinType) {
	this->pinType = pinType;
}

uint8_t ArduinoStatus::getPin(void) {
	return pin;
}

void ArduinoStatus::setPin(uint8_t pin) {
	this->pin = pin;
}

uint16_t ArduinoStatus::getPinValue(void) {
	return pinValue;
}

void ArduinoStatus::setPinValue(uint16_t pinValue) {
	this->pinValue = pinValue;
}

} /* namespace SISBARC */

File: maven-projects/java/jmx/arduino-base/src/main/java/br/com/cams7/arduino/util/ArduinoStatus.java
/**
 * 
 */
package br.com.cams7.arduino.util;

import java.util.Properties;

import br.com.cams7.util.AppException;
import br.com.cams7.util.AppUtil;

/**
 * @author cams7
 *
 */
public class ArduinoStatus {
	// Numero maximo da porta e 63
	public static final byte PIN_MAX = 0x3F;
	// Valor maximo da porta e 1023
	public static final short PIN_VALUE_MAX = 0x03FF;

	private static byte[] pinsDigital;
	private static byte[] pinsDigitalPWM;
	private static byte[] pinsAnalog;

	static {
		try {
			Properties pins = AppUtil.getPropertiesFile(ArduinoStatus.class,
					"pins.properties");
			final String SEPARATE = ",";

			pinsDigital = getPins(pins.getProperty("PIN_DIGITAL").trim()
					.split(SEPARATE));
			pinsDigitalPWM = getPins(pins.getProperty("PIN_DIGITAL_PWM").trim()
					.split(SEPARATE));
			pinsAnalog = getPins(pins.getProperty("PIN_ANALOG").trim()
					.split(SEPARATE));
		} catch (AppException e) {
			e.printStackTrace();
		}
	}

	private Transmitter transmitter;
	private Status status;
	private PinType pinType;

	private byte pin;
	private short pinValue;

	public ArduinoStatus() {
		super();

		setTransmitter(Transmitter.PC);
	}

	public ArduinoStatus(byte pin, short pinValue, Status status) {
		this();

		setStatus(status);

		setPin(pin);
		setPinValue(pinValue);

	}

	public void changeCurrentValues(ArduinoStatus arduino) {
		setTransmitter(arduino.getTransmitter());
		setStatus(arduino.getStatus());

		setPinType(arduino.getPinType());
		setPin(arduino.getPin());
		setPinValue(arduino.getPinValue());
	}

	private static byte[] getPins(final String[] values) {
		byte[] pins = new byte[values.length];
		for (byte i = 0; i < values.length; i++)
			pins[i] = Byte.parseByte(values[i].trim());
		return pins;
	}

	@Override
	public String toString() {
		return this.getClass().getSimpleName() + "[transmitter = "
				+ getTransmitter() + ", status = " + getStatus()
				+ ", pinType = " + getPinType() + ", pin = " + getPin()
				+ ", pinValue = " + getPinValue() + "]";
	}

	public Transmitter getTransmitter() {
		return transmitter;
	}

	public void setTransmitter(Transmitter transmitter) {
		this.transmitter = transmitter;
	}

	public Status getStatus() {
		return status;
	}

	public void setStatus(Status status) {
		this.status = status;
	}

	public PinType getPinType() {
		return pinType;
	}

	public void setPinType(PinType pinType) {
		this.pinType = pinType;
	}

	public byte getPin() {
		return pin;
	}

	public void setPin(byte pin) {
		this.pin = pin;
	}

	public short getPinValue() {
		return pinValue;
	}

	public void setPinValue(short pinValue) {
		this.pinValue = pinValue;
	}

	public enum Transmitter {
		ARDUINO, // Mensagem enviada do Arduino
		PC; // Mensagem enviada do PC
	}

	public enum Status {
		SEND, // Mensagem de envio que nao exige uma resposta
		SEND_RESPONSE, // Mensagem de envio que exige uma resposta
		RESPONSE, // Mensagem de resposta
		RESPONSE_RESPONSE;// Mensagem de resposta que exige outra resposta
	}

	public enum PinType {
		DIGITAL('d'), // Porta Digital
		ANALOG('a'); // Porta Analogica

		private char type;

		private PinType(char type) {
			this.type = type;
		}

		public char getType() {
			return type;
		}

	}

	public static byte[] getPinsDigital() {
		return pinsDigital;
	}

	public static byte[] getPinsDigitalPWM() {
		return pinsDigitalPWM;
	}

	public static byte[] getPinsAnalog() {
		return pinsAnalog;
	}

}
